executor/index_lookup_merge_join.go

```diff
@@ -0,0 +1,684 @@
+			task.done = true
+			if err != nil {
+				return errors.Trace(err)
+			}
+			task = e.getFinishedTask(ctx)
+		case <-ctx.Done():
+			breakFlag = true
+		}
+	}
+
+	return nil
+}
+
+func (e *IndexLookUpMergeJoin) getFinishedTask(ctx context.Context) *lookUpMergeJoinTask {
+	task := e.task
+	if task != nil && (!task.done || len(task.results) > 0) {
+		return task
+	}
+
+	select {
+	case task = <-e.resultCh:
+	case <-ctx.Done():
+		return nil
+	}
+	if task == nil {
+		return nil
+	}
+
+	if e.task != nil {
+		e.task.memTracker.Detach()
+	}
+	e.task = task
+	return task
+}
+
+func (omw *outerMergeWorker) run(ctx context.Context, wg *sync.WaitGroup) {
+	defer func() {
+		if r := recover(); r != nil {
+			err := errors.Errorf("%v", r)
+			logutil.Logger(contxt.Background()).Error("outerMergeWorker panic in the recoverable goroutine", zap.Error(err))
+			task := &lookUpMergeJoinTask{
+				results: make(chan *chunk.Chunk, 1),
+				doneErr: make(chan error, 1),
+				done:    true,
+			}
+			task.doneErr <- err
+			omw.pushToChan(ctx, task, omw.resultCh)
+		}
+		close(omw.resultCh)
+		close(omw.innerCh)
+		wg.Done()
+	}()
+	for {
+		task, err := omw.buildTask(ctx)
+		if err != nil {
+			task.doneErr <- errors.Trace(err)
+			omw.pushToChan(ctx, task, omw.resultCh)
+			return
+		}
+		if task == nil {
+			return
+		}
+
+		if finished := omw.pushToChan(ctx, task, omw.innerCh); finished {
+			return
+		}
+
+		if finished := omw.pushToChan(ctx, task, omw.resultCh); finished {
+			return
+		}
+	}
+}
+
+func (omw *outerMergeWorker) pushToChan(ctx context.Context, task *lookUpMergeJoinTask, dst chan<- *lookUpMergeJoinTask) bool {
+	select {
+	case <-ctx.Done():
+		return true
+	case dst <- task:
+	}
+	return false
+}
+
+// buildTask builds a lookUpMergeJoinTask and read outer rows.
+// When err is not nil, task must not be nil to send the error to the main thread via task
+func (omw *outerMergeWorker) buildTask(ctx context.Context) (*lookUpMergeJoinTask, error) {
+	omw.executor.newFirstChunk()
+
+	task := &lookUpMergeJoinTask{
+		results:     make(chan *chunk.Chunk, 1),
+		doneErr:     make(chan error, 1),
+		outerResult: omw.executor.newFirstChunk(),
+	}
+	task.memTracker = memory.NewTracker(fmt.Sprintf("lookup join task %p", task), -1)
+	task.memTracker.AttachTo(omw.parentMemTracker)
+
+	omw.increaseBatchSize()
+
+	task.memTracker.Consume(task.outerResult.MemoryUsage())
+	for task.outerResult.NumRows() < omw.batchSize {
+		err := omw.executor.Next(ctx, chunk.NewRecordBatch(omw.executorChk))
+		if err != nil {
+			return task, errors.Trace(err)
+		}
+		if omw.executorChk.NumRows() == 0 {
+			break
+		}
+
+		oldMemUsage := task.outerResult.MemoryUsage()
+		task.outerResult.Append(omw.executorChk, 0, omw.executorChk.NumRows())
+		newMemUsage := task.outerResult.MemoryUsage()
+		task.memTracker.Consume(newMemUsage - oldMemUsage)
+	}
+	if task.outerResult.NumRows() == 0 {
+		return nil, nil
+	}
+	if omw.outerMergeCtx.enforceOuterOrder == false {
```

Comment:
Seems like we need to reverse the meaning...

Reply:
There is a new change to this. PTAL in `executor/builder.go`.